crates/chain/src/local_chain.rs

```diff
@@ -1,14 +1,149 @@
+//! The [`LocalChain`] is a local implementation of [`ChainOracle`].
+
 use core::convert::Infallible;
 
-use alloc::collections::{BTreeMap, BTreeSet};
+use crate::collections::BTreeMap;
+use crate::{BlockId, ChainOracle};
+use alloc::sync::Arc;
 use bitcoin::BlockHash;
 
-use crate::{BlockId, ChainOracle};
+/// A structure that represents changes to [`LocalChain`].
+pub type ChangeSet = BTreeMap<u32, Option<BlockHash>>;
+
+/// Represents a block of [`LocalChain`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct CheckPoint(Arc<CPInner>);
+
+/// The internal contents of [`CheckPoint`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct CPInner {
+    /// Block id (hash and height).
+    block: BlockId,
+    /// Previous checkpoint (if any).
+    prev: Option<Arc<CPInner>>,
+}
+
+/// Occurs when the caller contructs a [`CheckPoint`] with a height that is not higher than the
+/// previous checkpoint it points to.
+#[derive(Debug, Clone, PartialEq)]
+pub struct NewCheckPointError {
+    /// The height of the new checkpoint.
+    pub new_height: u32,
+    /// The height of the previous checkpoint.
+    pub prev_height: u32,
+}
+
+impl core::fmt::Display for NewCheckPointError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "cannot construct checkpoint with a height ({}) that is not higher than the previous checkpoint ({})", self.new_height, self.prev_height)
+    }
+}
+
+impl std::error::Error for NewCheckPointError {}
+
+impl CheckPoint {
+    /// Construct a [`CheckPoint`] from a [`BlockId`].
+    pub fn new(block: BlockId) -> Self {
+        Self(Arc::new(CPInner { block, prev: None }))
+    }
+
+    /// Construct a [`CheckPoint`] of `block` with a previous checkpoint.
+    pub fn new_with_prev(
```

Comment:
Do we need this as a public API?

What about if you can just create a `CheckPoint` from a `BTreeSet` of `BlockId` which are assumed to be in the same chain.

Reply:
Ideally we want to create with a previous pointer. This way updates from the data source can connect in memory.

Reply:
What about a method on checkpoint like extend which creates a new one on top of an existing one?

Reply:
Heights may not increment in sequence, so don't we still need an error? We can also have a Result, where the error variant is the old checkpoint (meaning failed to extend).

Reply:
Yes sorry this wouldn't remove the error case but just make it a more natural API (pushing onto the end of a list).

Reply:
Changed the API to what you suggested @LLFourn